src/components/sections/contest/GithubFileSelect.tsx

```diff
@@ -0,0 +1,122 @@
+'use client'
+
+import {useState} from 'react'
+import {ArrowLeft} from 'lucide-react'
+import {signIn, useSession} from 'next-auth/react'
+
+import {Button} from '@/components/ui/Button'
+import {
+  useGetPublicRepos,
+  useGetRepoFiles,
+} from '@/lib/queries/github/getGitHub'
+import {
+  GetPublicReposResponse,
+  GetRepoFilesParams,
+  GetRepoFilesResponse,
+} from '@/server/actions/github/getGithub'
+import {Checkbox} from '@/components/ui/Checkbox'
+import {ScrollArea} from '@/components/ui/ScrollArea'
+
+const GithubFileSelect = () => {
+  const session = useSession()
+
+  const [selectedRepo, setSelectedRepo] = useState<GetRepoFilesParams | null>(
+    null,
+  )
+  const {data: publicReposData, isLoading: publicReposIsLoading} =
+    useGetPublicRepos()
+  const {data: repoFilesData, isLoading: repoFilesIsLoading} = useGetRepoFiles(
+    selectedRepo ?? {defaultBranch: '', owner: '', repo: ''},
+    {enabled: selectedRepo !== null},
+  )
+
+  if (session.data?.user.provider !== 'github') {
+    return (
+      <Button
+        onClick={() => signIn('github')}
+        className="bg-gray-800 text-white">
+        Sign in with Github to select repo files
+      </Button>
+    )
+  }
+
+  return (
+    <div className="min-w-96 rounded-md border-2 border-gray-900 p-2">
+      <div className="pb-2">
+        {selectedRepo === null ? (
+          <span className="font-bold">Select a repo</span>
+        ) : (
+          <div>
+            <Button variant="outline" onClick={() => setSelectedRepo(null)}>
+              <ArrowLeft className="h-4 w-4" />
+            </Button>
+            <span className="pl-2 font-bold">{`${selectedRepo.owner}/${selectedRepo.repo}`}</span>
+          </div>
+        )}
+      </div>
+      <ScrollArea className="h-96 w-full" thumbClassname="bg-gray-400">
+        <div className="flex flex-col items-start gap-2">
+          {selectedRepo === null ? (
+            <RepoList
+              repos={publicReposData}
+              isLoading={publicReposIsLoading}
+              selectRepo={(params: GetRepoFilesParams) =>
+                setSelectedRepo(params)
+              }
+            />
+          ) : (
+            <FileTree fileTree={repoFilesData} isLoading={repoFilesIsLoading} />
+          )}
+        </div>
+      </ScrollArea>
+    </div>
+  )
+}
+
+type RepoListProps = {
+  repos: GetPublicReposResponse | undefined
+  isLoading: boolean
+  selectRepo: (params: GetRepoFilesParams) => void
+}
```

Comment:
Just a tiny detail but considering the responsibilities of these components, `RepoList` shouldn't really care that `GithubFileSelect` needs `selectedRepo` formatted as `GetRepoFilesParams`. I think that `selectedRepo` should be of type `GetPublicReposResponse` and converting it to `GetRepoFilesParams` should happen when calling `useGetRepoFiles` in the parent.

Reply:
This came to me when I separated the repo and file select into their own components. Now this conversion needs to be handled by the parent since the components are controlled now